File: examples/POE2BP/POE2BP.c
// POE2BP.c, https://github.com/bhanson10/gbees/tree/main/examples/POE2BP
// Copyright 2025 by Benjamin Hanson, published under BSD 3-Clause License.

#include "../../gbees.h" 
#include "POE2BP.h"

// This function defines the dynamics model - required
void POE2BP(double* f, double* x, double t, double* coef){
    f[0] = 0;
    f[1] = pow(coef[0], 2.0) / pow(x[0], 3.0); 
}

int main(void){
    //=================================== Read in initial discrete measurement =================================//
    printf("Reading in initial discrete measurement...\n\n");

    char* P_DIR = "./results/c";       // Saved PDFs path
    char* M_DIR = "./measurements";    // Measurement path
    char* M_FILE = "measurement0.txt"; // Measurement file
    Meas M = Meas_create(DIM_f, M_DIR, M_FILE);
    //==========================================================================================================//

    //========================================== Read in user inputs ===========================================//
    printf("Reading in user inputs...\n\n");

    double factor[DIM_f] = {4.0, 2.5};
    Grid G = Grid_create(DIM_f, 0.0, 1E-9, M, factor); // Inputs: (dimension, initial time, probability threshold, measure, grid width factor)       

    double coef[] = {19.910350621818949};          // POE2BP trajectory attributes (mu)
    Traj T = Traj_create(1, coef);                 // Inputs: (# of coefficients, coefficients)

    int NUM_DIST = 5;                              // Number of distributions recorded per measurement
    int NUM_MEAS = 1;                              // Number of measurements
    int DEL_STEP = 20;                             // Number of steps per deletion procedure
    int OUTPUT_FREQ = 20;                          // Number of steps per output to terminal
    int CAPACITY = (int)pow(2,13);                 // Size of hash table (power of 2 for optimal hashing)
    bool OUTPUT = false;                           // Write info to terminal
    bool RECORD = true;                            // Write PDFs to .txt file
    bool MEASURE = false;                          // Take discrete measurement updates
    bool BOUNDS = false;                           // Add inadmissible regions to grid
    bool COLLISIONS = false;                       // Track collisions
    bool TV = false;                               // Time-invariant dynamics 
    //==========================================================================================================//

    //================================================= GBEES ==================================================//
    run_gbees(POE2BP, NULL, NULL, G, M, T, P_DIR, M_DIR, NUM_DIST, NUM_MEAS, DEL_STEP, OUTPUT_FREQ, CAPACITY, DIM_f, OUTPUT, RECORD, MEASURE, BOUNDS, COLLISIONS, TV);

    return 0;
}

File: examples/POE2BP/POE2BP.py
# POE2BP.py, https://github.com/bhanson10/gbees/tree/main/examples/POE2BP
# Copyright 2025 by Benjamin Hanson, published under BSD 3-Clause License.

import sys
sys.path.append('../../')
import gbeespy as gbees # type: ignore

DIM_f = 2 # State dimension

# This function defines the dynamics model - required
def POE2BP(x, t, coef):
    f0 = 0
    f1 = coef[0]**2.0 / x[0]**3.0; 
    return [f0, f1]

#==================================== Read in initial discrete measurement ==================================#
print("Reading in initial discrete measurement...\n")

P_DIR = "./results/python"      # Saved PDFs path
M_DIR = "./measurements"     # Measurement path
M_FILE = "measurement0.txt"  # Measurement file
M = gbees.Meas_create(DIM_f, M_DIR, M_FILE)
#============================================================================================================#

#=========================================== Read in user inputs ============================================#
print("Reading in user inputs...\n")

factor = [4.0, 2.5]
G = gbees.Grid_create(DIM_f, 0.0, 1E-9, M, factor, False) # Inputs: (dimension, initial time, probability threshold, measure, grid width factor, rotate grid)    

coef = [19.910350621818949]                               # POE2BP trajectory attributes (mu)
T = gbees.Traj_create(len(coef), coef)                    # Inputs: (# of coefficients, coefficients)

NUM_DIST = 5                                              # Number of distributions recorded per measurement
NUM_MEAS = 1                                              # Number of measurements
DEL_STEP = 20                                             # Number of steps per deletion procedure
OUTPUT_FREQ = 20                                          # Number of steps per output to terminal
CAPACITY = int(2**13);                                    # Size of hash table (power of 2 for optimal hashing)
OUTPUT = False                                            # Write info to terminal
RECORD = True                                             # Write PDFs to .txt file
MEASURE = False                                           # Take discrete measurement updates
BOUNDS = False                                            # Add inadmissible regions to grid
COLLISIONS = False;                                       # Track collisions
TV = False;                                               # Time-invariant dynamics 
BINARY = True;                                            # Binary output file
#============================================================================================================#

#================================================== GBEES ===================================================#
gbees.run_gbees(POE2BP, None, None, G, M, T, P_DIR, M_DIR, NUM_DIST, NUM_MEAS, DEL_STEP, OUTPUT_FREQ, CAPACITY, DIM_f, OUTPUT, RECORD, MEASURE, BOUNDS, COLLISIONS, TV, BINARY)